recipe-client-addon/lib/PreferenceExperiments.jsm

```diff
@@ -0,0 +1,296 @@
+    Preferences.observe(preferenceName, observerInfo.observer);
+  },
+
+  /**
+   * Disable a preference observer for the named experiment.
+   * @param {string} experimentName
+   * @throws {Error}
+   *   If there is no active observer for the named experiment.
+   */
+  stopObserver(experimentName) {
+    log.debug(`PreferenceExperiments.stopObserver(${experimentName})`);
+
+    if (!(experimentName in experimentObservers)) {
+      throw new Error(`No observer for the preference experiment ${experimentName} found.`);
+    }
+
+    const {preferenceName, observer} = experimentObservers[experimentName];
+    Preferences.ignore(preferenceName, observer);
+    delete experimentObservers[experimentName];
+  },
+
+  /**
+   * Disable all currently-active preference observers for experiments.
+   */
+  stopAllObservers() {
+    log.debug("PreferenceExperiments.stopAllObservers()");
+    for (const {preferenceName, observer} of Object.values(experimentObservers)) {
+      Preferences.ignore(preferenceName, observer);
+    }
+    experimentObservers = {};
+  },
+
+  /**
+   * Update the timestamp storing when Normandy last sent a recipe for the named
+   * experiment.
+   * @param {string} experimentName
+   * @rejects {Error}
+   *   If there is no stored experiment with the given name.
+   */
+  async markLastSeen(experimentName) {
+    log.debug(`PreferenceExperiments.markLastSeen(${experimentName})`);
+
+    const store = await loadStorage();
+    if (!(experimentName in store.data)) {
+      throw new Error(`Could not find a preference experiment named "${experimentName}"`);
+    }
+
+    store.data[experimentName].lastSeen = new Date().toJSON();
+    store.saveSoon();
+  },
+
+  /**
+   * Stop an active experiment, deactivate preference watchers, and optionally
+   * reset the associated preference to its previous value.
+   * @param {string} experimentName
+   * @param {boolean} [resetValue=true]
+   *   If true, reset the preference to its original value.
+   * @rejects {Error}
+   *   If there is no stored experiment with the given name, or if the
+   *   experiment has already expired.
+   */
+  async stop(experimentName, resetValue = true) {
+    log.debug(`PreferenceExperiments.stop(${experimentName})`);
+
+    const store = await loadStorage();
+    if (!(experimentName in store.data)) {
+      throw new Error(`Could not find a preference experiment named "${experimentName}"`);
+    }
+
+    const experiment = store.data[experimentName];
+    if (experiment.expired) {
+      throw new Error(
+        `Cannot stop preference experiment "${experimentName}" because it is already expired`
+      );
+    }
+
+    PreferenceExperiments.stopObserver(experimentName);
+
+    if (resetValue) {
+      const {preferenceName, previousPreferenceValue} = experiment;
+      Preferences.reset(preferenceName);
+      if (previousPreferenceValue !== undefined) {
+        Preferences.set(preferenceName, previousPreferenceValue);
```

Comment:
Reset will remove a preference that doesn't have a default value. This lets us support removing a preference from about:config entirely if it didn't exist there before, with `undefined` being the difference between an empty default value, like an empty string, and a preference that wasn't even there.